### ISSUE TITLE
Yarn PNP and TypeScript doesn't compile out of the box

### ISSUE THREAD

**Issue by alamothe** [facebook/docusaurus#6157]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [X] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

Yarn PNP and TypeScript doesn't compile out of the box. I had to add the following entries to `.yarnrc.yml` to make it work:

<details>

<summary><code>.yarnrc.yml</code></summary>

```yaml
  "@docusaurus/module-type-aliases@*":
    dependencies:
      "@types/history": "*"
      "@types/react-loadable": "*"
  "@docusaurus/plugin-content-blog@*":
    peerDependencies:
      "@docusaurus/types": "*"
  "@docusaurus/plugin-content-pages@*":
    peerDependencies:
      "@docusaurus/types": "*"
  "@docusaurus/plugin-content-docs@*":
    peerDependencies:
      "@docusaurus/types": "*"
    dependencies:
      "@docusaurus/module-type-aliases": "2.0.0-beta.14"
  "@docusaurus/theme-common@*":
    peerDependencies:
      "@docusaurus/types": "*"
      "@types/react": "*"
  "@docusaurus/theme-classic@*":
    peerDependencies:
      "@docusaurus/types": "*"
    dependencies:
      "utility-types": "*"
  "@docusaurus/preset-classic@*":
    peerDependencies:
      "@docusaurus/types": "*"
    dependencies:
      "@docusaurus/theme-common": "2.0.0-beta.14"
  "@docusaurus/mdx-loader@*":
    dependencies:
      "unified": "*"
  "@docusaurus/utils@*":
    peerDependencies:
      "@docusaurus/types": "*"
```

</details>

These are all dependencies that are missing or not properly declared.

### Steps to reproduce

1. `yarn set version berry`
2. `npx create-docusaurus@latest carsync-help classic --typescript`
3. `yarn tsc`

### Expected behavior

Compiles without errors.

### Actual behavior

<details>

<summary>Compilation errors reported:</summary>

```
../.yarn/__virtual__/@docusaurus-mdx-loader-virtual-adf7a41f60/0/cache/@docusaurus-mdx-loader-npm-2.0.0-beta.14-971ea241f9-248a8f73b4.zip/node_modules/@docusaurus/mdx-loader/src/mdx-loader.d.ts:8:27 - error TS2307: Cannot find module 'unified' or its corresponding type declarations.

8 import type {Plugin} from 'unified';
                            ~~~~~~~~~

../.yarn/__virtual__/@docusaurus-plugin-content-blog-virtual-e96af830a4/0/cache/@docusaurus-plugin-content-blog-npm-2.0.0-beta.14-3465a1b1cb-9d98707faa.zip/node_modules/@docusaurus/plugin-content-blog/src/plugin-content-blog.d.ts:29:30 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

29   import type {TOCItem} from '@docusaurus/types';
                                ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-plugin-content-docs-virtual-f0cdf7992e/0/cache/@docusaurus-plugin-content-docs-npm-2.0.0-beta.14-5f7a4ed55b-d7c1951f18.zip/node_modules/@docusaurus/plugin-content-docs/src/plugin-content-docs.d.ts:81:30 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

81   import type {TOCItem} from '@docusaurus/types';
                                ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-plugin-content-docs-virtual-f0cdf7992e/0/cache/@docusaurus-plugin-content-docs-npm-2.0.0-beta.14-5f7a4ed55b-d7c1951f18.zip/node_modules/@docusaurus/plugin-content-docs/src/types.ts:8:23 - error TS2688: Cannot find type definition file for '@docusaurus/module-type-aliases'.

8 /// <reference types="@docusaurus/module-type-aliases" />
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-plugin-content-pages-virtual-dbf42224e5/0/cache/@docusaurus-plugin-content-pages-npm-2.0.0-beta.14-0bbd113bd5-1372f554d8.zip/node_modules/@docusaurus/plugin-content-pages/src/plugin-content-pages.d.ts:13:30 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

13   import type {TOCItem} from '@docusaurus/types';
                                ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-preset-classic-virtual-ab54b4a240/0/cache/@docusaurus-preset-classic-npm-2.0.0-beta.14-c882f4954f-67019820c8.zip/node_modules/@docusaurus/preset-classic/src/preset-classic.d.ts:19:34 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

19 export type ThemeConfig = import('@docusaurus/types').ThemeConfig &
                                    ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-preset-classic-virtual-ab54b4a240/0/cache/@docusaurus-preset-classic-npm-2.0.0-beta.14-c882f4954f-67019820c8.zip/node_modules/@docusaurus/preset-classic/src/preset-classic.d.ts:20:10 - error TS2307: Cannot find module '@docusaurus/theme-common' or its corresponding type declarations.

20   import('@docusaurus/theme-common').UserThemeConfig & {
            ~~~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-classic-virtual-ec17751a11/0/cache/@docusaurus-theme-classic-npm-2.0.0-beta.14-ecbe2a9dfe-ebb4c362e4.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:602:30 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

602   import type {TOCItem} from '@docusaurus/types';
                                 ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-classic-virtual-ec17751a11/0/cache/@docusaurus-theme-classic-npm-2.0.0-beta.14-ecbe2a9dfe-ebb4c362e4.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:617:30 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

617   import type {TOCItem} from '@docusaurus/types';
                                 ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-classic-virtual-ec17751a11/0/cache/@docusaurus-theme-classic-npm-2.0.0-beta.14-ecbe2a9dfe-ebb4c362e4.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:642:30 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

642   import type {TOCItem} from '@docusaurus/types';
                                 ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-classic-virtual-ec17751a11/0/cache/@docusaurus-theme-classic-npm-2.0.0-beta.14-ecbe2a9dfe-ebb4c362e4.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:655:30 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

655   import type {TOCItem} from '@docusaurus/types';
                                 ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-classic-virtual-ec17751a11/0/cache/@docusaurus-theme-classic-npm-2.0.0-beta.14-ecbe2a9dfe-ebb4c362e4.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:809:31 - error TS2307: Cannot find module 'utility-types' or its corresponding type declarations.

809   import type {Optional} from 'utility-types';
                                  ~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-common-virtual-58b8d3e7b7/0/cache/@docusaurus-theme-common-npm-2.0.0-beta.14-de796828fc-dbe969d676.zip/node_modules/@docusaurus/theme-common/lib/utils/historyUtils.d.ts:7:15 - error TS2305: Module '"@docusaurus/history"' has no exported member 'Location'.

7 import type { Location, Action } from '@docusaurus/history';
                ~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-common-virtual-58b8d3e7b7/0/cache/@docusaurus-theme-common-npm-2.0.0-beta.14-de796828fc-dbe969d676.zip/node_modules/@docusaurus/theme-common/lib/utils/historyUtils.d.ts:7:25 - error TS2305: Module '"@docusaurus/history"' has no exported member 'Action'.

7 import type { Location, Action } from '@docusaurus/history';
                          ~~~~~~

../.yarn/__virtual__/@docusaurus-theme-common-virtual-58b8d3e7b7/0/cache/@docusaurus-theme-common-npm-2.0.0-beta.14-de796828fc-dbe969d676.zip/node_modules/@docusaurus/theme-common/lib/utils/tocUtils.d.ts:7:25 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

7 import { TOCItem } from '@docusaurus/types';
                          ~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-common-virtual-58b8d3e7b7/0/cache/@docusaurus-theme-common-npm-2.0.0-beta.14-de796828fc-dbe969d676.zip/node_modules/@docusaurus/theme-common/lib/utils/useLocationChange.d.ts:7:10 - error TS2305: Module '"@docusaurus/history"' has no exported member 'Location'.

7 import { Location } from '@docusaurus/history';
           ~~~~~~~~

../.yarn/__virtual__/@docusaurus-utils-virtual-064bf14a46/0/cache/@docusaurus-utils-npm-2.0.0-beta.14-d4176db19a-f527eeeca5.zip/node_modules/@docusaurus/utils/lib/index.d.ts:7:76 - error TS2307: Cannot find module '@docusaurus/types' or its corresponding type declarations.

7 import { ReportingSeverity, TranslationFileContent, TranslationFile } from '@docusaurus/types';
                                                                             ~~~~~~~~~~~~~~~~~~~

../.yarn/cache/@docusaurus-module-type-aliases-npm-2.0.0-beta.14-2a8c7bc4e7-d4bb46da11.zip/node_modules/@docusaurus/module-type-aliases/src/index.d.ts:99:44 - error TS2307: Cannot find module 'react-loadable' or its corresponding type declarations.

99   import type {LoadingComponentProps} from 'react-loadable';
                                              ~~~~~~~~~~~~~~~~

../.yarn/cache/@docusaurus-module-type-aliases-npm-2.0.0-beta.14-2a8c7bc4e7-d4bb46da11.zip/node_modules/@docusaurus/module-type-aliases/src/index.d.ts:233:17 - error TS2307: Cannot find module 'history' or its corresponding type declarations.

233   export * from 'history';
                    ~~~~~~~~~

../.yarn/cache/@docusaurus-module-type-aliases-npm-2.0.0-beta.14-2a8c7bc4e7-d4bb46da11.zip/node_modules/@docusaurus/module-type-aliases/src/index.d.ts:275:29 - error TS2307: Cannot find module 'react-loadable' or its corresponding type declarations.

275   import type Loadable from 'react-loadable';
                                ~~~~~~~~~~~~~~~~


Found 20 errors.
```

</details>

### Your environment

- Docusaurus version used: 2.0.0-beta.14

### Self-service

- [X] I'd be willing to fix this bug myself.

**Comment by Josh-Cena** [comment]
Yes, I also encountered this in my project (`@docusaurus/types` missing as dependency). However I also don't want to include these type-only packages as production dependency because I assume that's not the usual practice. @merceyz What would be a good middleground? Should we actually declare dependency on types packages because our exported types rely on them?

**Comment by arcanis** [comment]
They should be added in the template, as development dependency, since they are referenced in the generated config file.

Also note that I think there might be a mistake in the template, as I don't see a `Config` type in `@docusaurus/types`, referenced in the generated `docusaurus.config.js` (there is a `DocusaurusConfig` type, but it contains a few extra fields that the template doesn't list).

**Comment by Josh-Cena** [comment]
There's indeed a `Config` type here:

https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-types/src/index.d.ts#L78

This is the deeply partial version of `DocusaurusConfig` before any validation / normalization

**Comment by arcanis** [comment]
I see - I got the wrong version due to your versioning scheme: you published a `2.0.0-beta.ff31de0ff` in August, which per semver rules is higher than `^2.0.0-beta.14`, so that's the version which got installed 🤔

**Comment by Josh-Cena** [comment]
Haha, we have another person who got tricked... See also #6140. It seems to be only caused when migrating to PnP though? Time to clear up my doubt: does Yarn take timestamps into account when resolving versions?

**Comment by arcanis** [comment]
It's not PnP, just the regular Yarn resolution. We take all the versions, filter the compatible ones, and get the max satisfying one. Because semver is weird, it can be problematic with things like this ...

Other package managers perhaps don't sort the versions returned by the registry, but we do since semver resolution isn't supposed to be order-dependent 🤔

**Comment by Josh-Cena** [comment]
Yeah, we have ~150 beta canary version names that could be messing with semver, wondering if we should deprecate them all.

**Comment by alamothe** [comment]
Same here, I used `^2.0.0-beta.14` and got the wrong version, then almost posted on Yarn discord about it.

**Comment by alamothe** [comment]
Going back to the issue, I am of the opinion if you export types from a different package, it should be stated as a dependency, otherwise it's relying on that package being included by someone else. If not, it's essentially saying let's ignore the concept of `dependencies`/`peerDependencies` for types altogether. 

The downside is that if the user does not care about TypeScript at all, they would still get this dependency. Perhaps there should be a concept of `typeDependency`?

**Comment by Josh-Cena** [comment]
@arcanis Also, it seems like `"extends": "@tsconfig/docusaurus/tsconfig.json",` doesn't work in PnP? Does Yarn have custom resolver support for this?

**Comment by arcanis** [comment]
> The downside is that if the user does not care about TypeScript and uses plain JavaScript, they would still get this dependency. Perhaps there should be a concept of `typeDependency`?

There aren't (very short discussion [here](https://github.com/microsoft/DefinitelyTyped-tools/issues/43), there are a few other threads). At the moment the best option would be to use `peerDependencies`, and let package managers implement @types pruning.

> Also, it seems like `"extends": "@tsconfig/docusaurus/tsconfig.json"`, doesn't work in PnP? Does Yarn have custom resolver support for this?

@Josh-Cena Seems like we missed this spot when adding PnP support to TS, I'll make an issue on our side to address that.

**Comment by Josh-Cena** [comment]
Yeah, the reason why I deleted my comment🤦‍♂️Didn't bump my SDK version so the resolver was failing

**Comment by slorber** [comment]
Hey,

> I used ^2.0.0-beta.14 and got the wrong version

According to this tool https://semver.npmjs.com/ there's only one version that matches. Isn't Yarn supposed to implement the exact same rules? 🤷‍♂️

**Comment by Josh-Cena** [comment]
@slorber I can't figure out how that tool's supposed to be used... Here's what I tried:

![image](https://user-images.githubusercontent.com/55398995/147676463-85cc6a78-acb5-48dc-9d4f-057951aadc63.png)

**Comment by arcanis** [comment]
Iirc the npm semver calculator uses a very old version of the semver package, with some different results.

**Comment by slorber** [comment]
Yes, I'm not very confident with these range things.

Looking at how to deprecate all those legacy canary releases, not sure what's the best way. 

Executing npm deprecate commands one by one for all versions/packages is going to take forever. 

It takes a range, but will it actually work for our use-case? I might as well deprecate all betas and then undeprecate the few betas we want to keep 🤷‍♂️ or use the npm API directly

**Comment by Josh-Cena** [comment]
Just deprecate all versions greater than `beta.10000` then :D

**Comment by slorber** [comment]
This command seems to have worked:

```
npm deprecate @docusaurus/core@">= 2.0.0-beta" "deprecated canary release, please upgrade"

npm deprecate @docusaurus/core@"2.0.0-beta.0 || 2.0.0-beta.1 || 2.0.0-beta.2 || 2.0.0-beta.3 || 2.0.0-beta.4 || 2.0.0-beta.5 || 2.0.0-beta.6 || 2.0.0-beta.7 || 2.0.0-beta.8 || 2.0.0-beta.9 || 2.0.0-beta.10 || 2.0.0-beta.11 || 2.0.0-beta.12 || 2.0.0-beta.13 || 2.0.0-beta.14" ""
```

Tell me if this improves anything

Should I run this on all packages, or core would be enough?

**Comment by JaimeObregon** [comment]
I got tricked by this too 😅🔫. See cloud-annotations/docusaurus-openapi/issues/150.

**Comment by bourdakos1** [comment]
@slorber I'm not sure if deprecating is enough. It seems like `@docusaurus/core@npm:2.0.0-beta.fc64c12e4` is deprecated, but yarn 3.1.1 still installs it (we use `"^2.0.0-beta.14"` for all docusaurus package versions):

```
@docusaurus/core@npm:2.0.0-beta.fc64c12e4 is deprecated: deprecated canary release, please upgrade
```

**Comment by arcanis** [comment]
I'll check tomorrow, but we're supposed to prioritize non-deprecated releases if there's any matching the requested range (failing that, we pick a deprecated one) 🤔

**Comment by bourdakos1** [comment]
Nevermind, I think the deprecation works as expected. Installing `"@docusaurus/core": "^2.0.0-beta.14"` on its own, installs `2.0.0-beta.14`. 

However, when installing other docusaurus packages alongside `@docusaurus/core` (say `@docusaurus/preset-classic`), that will install `@docusaurus/preset-classic@2.0.0-beta.fc64c12e4`, which depends on `@docusaurus/core@2.0.0-beta.fc64c12e4`

@slorber So deprecating all the other old packages should fix the issue :)

**Comment by Josh-Cena** [comment]
> will install `@docusaurus/preset-classic@2.0.0-beta.fc64c12e4`, which depends on `@docusaurus/core@2.0.0-beta.fc64c12e4`

Does that happen to be because of bad lock file?

**Comment by bourdakos1** [comment]
No, I think this is the expected behavior

**Comment by slorber** [comment]
I see, so what was the solution to unlock this? Regenerate the whole lockfile?

**Comment by bourdakos1** [comment]
What worked for our particular scenario was adding a bunch of resolutions and re-running `yarn install`, no need to delete yarn.lock or anything:
```json
"resolutions": {
  "@docusaurus/core": "2.0.0-beta.14",
  "@docusaurus/preset-classic": "2.0.0-beta.14",
  "@docusaurus/mdx-loader": "2.0.0-beta.14",
  "@docusaurus/plugin-content-docs": "2.0.0-beta.14",
  "@docusaurus/types": "2.0.0-beta.14",
  "@docusaurus/utils-validation": "2.0.0-beta.14",
  "@docusaurus/utils": "2.0.0-beta.14"
}
```

I think deprecating the rest of the canary releases for all other Docusaurus packages would remove the need for the resolutions

**Comment by slorber** [comment]
Thanks @bourdakos1 

Just deprecated all the packages, let me know if you are able to remove those resolutions

Hope I didn't break anything 🤪

**Comment by bourdakos1** [comment]
It worked 🎉 Thanks for the help 😁

**Comment by Josh-Cena** [comment]
I've done some work in #7521 and fixed all missing dependency problems. However when running `tsc` I still get a few errors, which seem cryptic to me:

```
.yarn/__virtual__/@docusaurus-theme-classic-virtual-59c6f4f552/0/cache/@docusaurus-theme-classic-npm-0.0.0-5076-a80de9154f-0c8b55566c.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:218:35 - error TS2339: Property 'getLineProps' does not exist on type '{}'.

218   type GetLineProps = RenderProps['getLineProps'];
                                      ~~~~~~~~~~~~~~

.yarn/__virtual__/@docusaurus-theme-classic-virtual-59c6f4f552/0/cache/@docusaurus-theme-classic-npm-0.0.0-5076-a80de9154f-0c8b55566c.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:219:36 - error TS2339: Property 'getTokenProps' does not exist on type '{}'.

219   type GetTokenProps = RenderProps['getTokenProps'];
                                       ~~~~~~~~~~~~~~~

.yarn/__virtual__/@docusaurus-theme-classic-virtual-59c6f4f552/0/cache/@docusaurus-theme-classic-npm-0.0.0-5076-a80de9154f-0c8b55566c.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:220:28 - error TS2339: Property 'tokens' does not exist on type '{}'.

220   type Token = RenderProps['tokens'][number][number];
                               ~~~~~~~~

.yarn/cache/unified-npm-9.2.2-65676eec78-7c24461be7.zip/node_modules/unified/types/ts4.0/index.d.ts:3:20 - error TS2307: Cannot find module 'unist' or its corresponding type declarations.

3 import {Node} from 'unist'
                     ~~~~~~~
```

When I go into `.yarn/__virtual__/@docusaurus-theme-classic-virtual-59c6f4f552/0/cache/@docusaurus-theme-classic-npm-0.0.0-5076-a80de9154f-0c8b55566c.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts`, I can see that `RenderProps` has correct types, just that `tsc` cannot resolve it somehow.

**Comment by merceyz** [comment]
You can try running the following command and see if it gives you some more details.
```sh
PNP_DEBUG_LEVEL=1 yarn tsc
```

**Comment by alamothe** [comment]
This is what I'm getting with the latest version:
```bash
../.yarn/__virtual__/@docsearch-react-virtual-d0cb4410d8/0/cache/@docsearch-react-npm-3.2.0-6c2ef6865a-219c15259c.zip/node_modules/@docsearch/react/dist/esm/DocSearch.d.ts:2:36 - error TS2307: Cannot find module '@algolia/client-search' or its corresponding type declarations.

2 import type { SearchOptions } from '@algolia/client-search';
                                     ~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-classic-virtual-1475f1f307/0/cache/@docusaurus-theme-classic-npm-2.0.1-dbd0568bd1-f37077f394.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:1150:44 - error TS2307: Cannot find module '@docusaurus/theme-common/Details' or its corresponding type declarations.

1150   import {Details, type DetailsProps} from '@docusaurus/theme-common/Details';
                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-classic-virtual-1475f1f307/0/cache/@docusaurus-theme-classic-npm-2.0.1-dbd0568bd1-f37077f394.zip/node_modules/@docusaurus/theme-classic/src/theme-classic.d.ts:1172:34 - error TS2307: Cannot find module '@docusaurus/theme-common/internal' or its corresponding type declarations.

1172   import type {TOCTreeNode} from '@docusaurus/theme-common/internal';
                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-common-virtual-a8ce414b90/0/cache/@docusaurus-theme-common-npm-2.0.1-73566d28ef-65261e6f77.zip/node_modules/@docusaurus/theme-common/lib/contexts/docsPreferredVersion.d.ts:8:36 - error TS2307: Cannot find module '@docusaurus/plugin-content-docs/client' or its corresponding type declarations.

8 import { type GlobalVersion } from '@docusaurus/plugin-content-docs/client';
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/__virtual__/@docusaurus-theme-common-virtual-a8ce414b90/0/cache/@docusaurus-theme-common-npm-2.0.1-73566d28ef-65261e6f77.zip/node_modules/@docusaurus/theme-common/lib/utils/docsUtils.d.ts:8:72 - error TS2307: Cannot find module '@docusaurus/plugin-content-docs/client' or its corresponding type declarations.

8 import { type GlobalVersion, type GlobalSidebar, type GlobalDoc } from '@docusaurus/plugin-content-docs/client';
                                                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/cache/@algolia-autocomplete-core-npm-1.7.1-488e460895-511176e9c2.zip/node_modules/@algolia/autocomplete-core/dist/esm/types/AutocompleteSource.d.ts:1:43 - error TS2307: Cannot find module '@algolia/autocomplete-preset-algolia' or its corresponding type declarations.

1 import type { RequesterDescription } from '@algolia/autocomplete-preset-algolia';
                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/cache/@algolia-autocomplete-core-npm-1.7.1-488e460895-511176e9c2.zip/node_modules/@algolia/autocomplete-core/dist/esm/utils/mapToAlgoliaResponse.d.ts:1:67 - error TS2307: Cannot find module '@algolia/client-search' or its corresponding type declarations.

1 import type { SearchForFacetValuesResponse, SearchResponse } from '@algolia/client-search';
                                                                    ~~~~~~~~~~~~~~~~~~~~~~~~

../.yarn/cache/@algolia-autocomplete-core-npm-1.7.1-488e460895-511176e9c2.zip/node_modules/@algolia/autocomplete-core/dist/esm/utils/mapToAlgoliaResponse.d.ts:4:18 - error TS2307: Cannot find module '@algolia/client-search' or its corresponding type declarations.

4     hits: import("@algolia/client-search").Hit<THit>[][];
                   ~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by Josh-Cena** [comment]
@alamothe No; please compile with `"moduleResolution": "NodeNext"`. We are using package.json `exports` field.

**Comment by slorber** [comment]
This is an old issue and we have been able to enable Yarn PnP in our CI e2e tests so that our init template can at least work using it: 

https://github.com/facebook/docusaurus/blob/main/.github/workflows/tests-e2e.yml#L75


It's working (and we'll try to keep it this way)
---

But to be honest I wouldn't recommend using Yarn PnP with Docusaurus. It might work for some, but whenever I have to troubleshoot an open-source site using it, it's really time-consuming for me to try to troubleshoot things that otherwise work with the node linker, and is a real time sink.

Just a few examples:
- Someone recently mentioned that our Mermaid theme doesn't work with PnP (not covered by our e2e tests): https://github.com/facebook/docusaurus/issues/10645
- PnP doesn't seem to work with Docusaurus Faster (in particular Rspack): https://github.com/easyops-cn/docusaurus-search-local/pull/468#issuecomment-2454715676

It's hard and time-consuming for me to ensure that all things work properly under both linkers, unfortunately. If you adopt PnP, please consider that you might have to figure out problems on your own. I can only commit to making a basic Docusaurus site work under PnP, but can't ensure that all third-party plugins, libs, and packages will work well using it.

**Comment by arcanis** [comment]
Of note, there's work on progress to integrate PnP with rspack; I'll report back when that changes.

**Comment by slorber** [comment]
Great news thanks!

I'd be happy if we could improve PnP support, but I lack good troubleshooting skills for Yarn PnP.

Since I can't inspect `node_modules` I find it hard to troubleshoot duplicate packages (for example `react` or `@docusaurus/theme-common` being used twice, leading to hooks errors).

I also struggle to understand how to deal with optional peer dependencies, which our v3.6 Docusaurus faster project rely on extensively. On a site upgrade using PnP (https://github.com/easyops-cn/docusaurus-search-local/pull/468) I figured out we had to declare a few extra dependencies for packages that are not always in our control and also using optional peer dependencies:

```yaml
packageExtensions:
  "@docusaurus/bundler@*":
    dependencies:
      "@docusaurus/faster": "*"
  "@docusaurus/theme-common@*":
    dependencies:
      "@docusaurus/plugin-content-docs": "*"
  "terser-webpack-plugin@*":
    dependencies:
      "@swc/core": "*"
  "css-minimizer-webpack-plugin@*":
    dependencies:
      "lightningcss": "*"
```
